StrongInject.Tests.Unit/OwnedInjectionTests.cs

```diff
@@ -0,0 +1,1950 @@
+            catch
+            {
+                ((global::System.IDisposable)c_0_1).Dispose();
+                throw;
+            }
+
+            return new global::StrongInject.Owned<global::B>(b_0_0, () =>
+            {
+                ((global::System.IDisposable)c_0_1).Dispose();
+            });
+        }
+
+        owned_0_2 = CreateOwnedB_3();
+        a_0_0 = new global::A(b1: func_0_1, b2: owned_0_2);
+        TResult result;
+        try
+        {
+            result = func(a_0_0, param);
+        }
+        finally
+        {
+        }
+
+        return result;
+    }
+
+    global::StrongInject.Owned<global::A> global::StrongInject.IContainer<global::A>.Resolve()
+    {
+        if (Disposed)
+            throw new global::System.ObjectDisposedException(nameof(Container));
+        global::System.Func<global::C, global::StrongInject.Owned<global::B>> func_0_1;
+        global::StrongInject.Owned<global::B> owned_0_2;
+        global::A a_0_0;
+        func_0_1 = (param0_0) =>
+        {
+            global::StrongInject.Owned<global::B> owned_1_0;
+            global::StrongInject.Owned<global::B> CreateOwnedB_1()
+            {
+                global::B b_1_0;
+                b_1_0 = new global::B(c: param0_0);
+                return new global::StrongInject.Owned<global::B>(b_1_0, () =>
+                {
+                });
+            }
+
+            owned_1_0 = CreateOwnedB_1();
+            return owned_1_0;
+        };
+        global::StrongInject.Owned<global::B> CreateOwnedB_3()
+        {
+            global::C c_0_1;
+            global::B b_0_0;
+            c_0_1 = new global::C();
+            try
+            {
+                b_0_0 = new global::B(c: c_0_1);
+            }
+            catch
+            {
+                ((global::System.IDisposable)c_0_1).Dispose();
+                throw;
+            }
+
+            return new global::StrongInject.Owned<global::B>(b_0_0, () =>
+            {
+                ((global::System.IDisposable)c_0_1).Dispose();
+            });
+        }
+
+        owned_0_2 = CreateOwnedB_3();
+        a_0_0 = new global::A(b1: func_0_1, b2: owned_0_2);
+        return new global::StrongInject.Owned<global::A>(a_0_0, () =>
+        {
+        });
+    }
+}");
+        }
+
+        [Fact]
+        public void AsyncOwnedIsRequiredForAsyncDisposables()
+        {
+            string userSource = @"
+using System;
+using System.Threading.Tasks;
+using StrongInject;
+
+[Register(typeof(A)), Register(typeof(B))]
+public partial class Container : IAsyncContainer<A> { }
+
+public record A(Func<Owned<B>> b);
+public record B : IAsyncDisposable { public ValueTask DisposeAsync() => default; }
+";
+            _ = RunGeneratorWithStrongInjectReference(userSource, out var generatorDiagnostics, out _);
+            generatorDiagnostics.Verify(
+                // (7,22): Warning SI1301: Cannot call asynchronous dispose for 'B' in implementation of synchronous container
```

Comment:
Is this message confusing? It's a synchronous Owned type in an asynchronous container.

Reply:
I think it is - if possible can we thread that this is an Owned through and give a separate diagnostic?

Reply:
Done.